### PR TITLE
DAOS-11977 cart: Reorganize provider tables and initialization

### PR DESCRIPTION
**Pull request by frostedcmos**
Reorganize provider tables and initialization

Signed-off-by: Alexander A Oganezov <alexander.a.oganezov@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by jgmoore-or — APPROVED** [review]

File: src/cart/crt_init.c [hunk 1]
```diff
@@ -23,13 +23,15 @@ dump_envariables(void)
 {
 	int	i;
 	char	*val;
-	char	*envars[] = {"CRT_PHY_ADDR_STR", "D_LOG_STDERR_IN_LOG",
+	char	*envars[] = {"D_PROVIDER", "D_INTERFACE", "D_DOMAIN", "D_PORT",
```
Comment (soumagne) [inline src/cart/crt_init.c:26]: Why do we have more env vars now ? they seem to be duplicate of previous OFI_PORT etc ?

Comment (frostedcmos) [reply src/cart/crt_init.c:26]: @johannlombardi wanted to standardize envariables as names like CRT_PHY_ADDR_STR are more confusing

**Review by soumagne — APPROVED** [review]

File: src/cart/crt_context.c [hunk 2]
```diff
@@ -173,7 +173,42 @@ crt_context_init(crt_context_t crt_ctx)
...
+
+	if (crt_ctx == NULL || uri == NULL) {
+		D_ERROR("Invalid null parameters\n");
+		return -DER_INVAL;
+	}
+
+	ctx = crt_ctx;
+	D_STRNDUP(*uri, ctx->cc_self_uri, CRT_ADDR_STR_MAX_LEN);
+	if (*uri == NULL)
+		return DER_NOMEM;
```
Comment (ashleypittman) [inline src/cart/crt_context.c:205]: @frostedcmos shouldn't this be -DER_NOMEM?

Comment (frostedcmos) [reply src/cart/crt_context.c:205]: yep good catch. will patch up in the next pr

**Pull request merged**